### ISSUE TITLE
We're not warning against fn ptr and wide raw ptr nested inside ADTs

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#116929]
I found a gap in our match forward-compatibility warnings:
```rust
type T = (*const [i32], i32);
const C: T = (unsafe { std::mem::transmute((0usize, 0usize)) }, 0);

fn test(x: T) -> bool {
    matches!(x, C)
}
```
This should warn but doesn't.

The reason is that [here](https://github.com/rust-lang/rust/blob/223674a824317284b42ba197be43fd68b957ab58/compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs#L239-L252) we are only checking the top-level type to be a wide raw ptr or fn ptr, we are not recursing.

The fix IMO is to reject those types in `compiler/rustc_trait_selection/src/traits/structural_match.rs`, which already recurses the type.

Cc @oli-obk @lcnr

**Comment by oli-obk** [comment]
> The fix IMO is to reject those types in `compiler/rustc_trait_selection/src/traits/structural_match.rs`, which already recurses the type.

That would warn on `Option<String>`, which can be used at None value

**Comment by RalfJung** [comment]
> That would warn on Option<String>, which can be used at None value

It would? That's surprising, since valtree construction will succeed for `Option<String>`.
So that could be used as a signal not to lint.
